xapian-core/cluster/lcd_clusterer.cc

```diff
@@ -0,0 +1,164 @@
+/** @file lcd_clusterer.cc
+ *  @brief LCD clustering API
+ */
+/* Copyright (C) 2018 Uppinder Chugh
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian/cluster.h"
+#include "xapian/error.h"
+
+#include "debuglog.h"
+
+#include <algorithm>
+#include <vector>
+
+using namespace Xapian;
+using namespace std;
+
+typedef set<pair<Point, double> > PSet;
+
+LCDClusterer::LCDClusterer(unsigned int k_)
+    : k(k_)
+{
+    LOGCALL_CTOR(API, "LCDClusterer", k_);
+    if (k_ == 0)
+	throw InvalidArgumentError("Number of required clusters should be "
+				   "greater than zero");
+}
+
+string
+LCDClusterer::get_description() const
+{
+    return "LCDClusterer()";
+}
+
+struct pcompare {
+    bool operator() (const pair<Point, double> &a,
+		     const pair<Point, double> &b) const {
+	return a.second > b.second;
+    }
+};
+
+struct dcompare {
+    bool operator() (const pair<PSet::iterator, double> &a,
+		     const pair<PSet::iterator, double> &b) const {
+	return a.second < b.second;
+    }
+};
+
+ClusterSet
+LCDClusterer::cluster(const MSet &mset)
+{
+    LOGCALL(API, ClusterSet, "LCDClusterer::cluster", mset);
+
+    doccount size = mset.size();
+    if (k >= size)
+	k = size;
+
+    // Store each document and it's rel score from given mset
```

Comment:
Not a contraction of "it is", so should be "its".

Reply:
Modification pushed in a fixup commit.